strawberryfields/tdm/tdmprogram.py

```diff
@@ -0,0 +1,327 @@
+
+    If we print out this program, we see that the time domain program
+    has automated the process of repeating the single time-bin sequence
+    constructed above:
+
+    >>> prog.print()
+    Sgate(0.32, 0) | (q[1])
+    BSgate(1, 0) | (q[0], q[1])
+    MeasureHomodyne(3) | (q[0])
+    Sgate(0.32, 0) | (q[0])
+    BSgate(2, 0) | (q[1], q[0])
+    MeasureHomodyne(4) | (q[1])
+    Sgate(0.32, 0) | (q[1])
+    BSgate(1, 0) | (q[0], q[1])
+    MeasureHomodyne(3) | (q[0])
+    Sgate(0.32, 0) | (q[0])
+    BSgate(2, 0) | (q[1], q[0])
+    MeasureHomodyne(4) | (q[1])
+    Sgate(0.32, 0) | (q[1])
+    BSgate(1, 0) | (q[0], q[1])
+    MeasureHomodyne(3) | (q[0])
+    Sgate(0.32, 0) | (q[0])
+    BSgate(2, 0) | (q[1], q[0])
+    MeasureHomodyne(4) | (q[1])
+
+    Note the 'shifting' of the measured registers --- this optimization allows
+    for time domain algorithms to be simulated in memory much more efficiently:
+
+    >>> eng = sf.Engine("gaussian")
+    >>> results = eng.run(prog)
+
+    The engine automatically takes this mode shifting into account; returned samples
+    will always be transformed to match the modes specified during construction:
+
+    >>> print(results.all_samples)
+    {0: [array([1.26208025]), array([1.53910032]), array([-1.29648336]),
+    array([0.75743215]), array([-0.17850101]), array([-1.44751996])]}
+
+    Note that, unlike the standard :class:`~.Program`,
+    the TDM context manager has both required and essential arguments:
+
+    * Positional arguments are used to pass sequences of gate arguments
+      to apply per time-bin. Within the context, these are accessible via the ``p``
+      context variable; ``p[i]`` corresponds to the ``i``-th positional
+      arguments.
+
+      All positional arguments corresponding to sequences of gate arguments
+      **must be the same length**. This length determines how many time-bins
+      are present in the TDM program.
+
+      If the ``p`` variable is not used, the gate argument is assumed to be **constant**
+      across all time-bins.
+
+    * ``copies=1`` *(int)*: the number of times to repeat the time-domain program.
+      For example, if a sequence of 10 gate arguments is provided, and ``copies==15``,
+      then there will be a total of :math:`10\times 15` time bins in the time domain
+      program. The sequence corresponding to the gate arguments is repeated ``copies``
+      number of times.
+
+    * ``shift="default"`` *(str or int)*: defines in which way the qumode register is shifted at the end of each time bin. If set to "default", the qumode register is shifted such that each measured qumode reappears as a fresh mode at the beginning of the subsequent time bin. This is equivalent to a measured qumode being removed from the representation (by measurement) and a new one being added (by a light source). If set to an integer value, the register will shift by a step size of this integer at the end of each time bin.
```

Comment:
Feel free to modify as you wish. Question: Always wondered how you set the word wrap column in these strings?

Reply:
I have an extension in my text editor that automatically wraps to 100 characters :)